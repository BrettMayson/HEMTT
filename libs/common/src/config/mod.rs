#![allow(clippy::module_name_repetitions)]

//! Read the project configuration into a [`ProjectConfig`] struct

use std::path::Path;

use tracing::warn;

mod addon;
mod pdrive;
mod project;

pub use addon::AddonConfig;
pub use pdrive::PDriveOption;
pub use project::{hemtt::launch::LaunchOptions, ProjectConfig};

fn deprecated(file: &Path, key: &str, replacement: &str, info: Option<&str>) {
    warn!(
        "Use of deprecated key '{}' in '{}'. Use '{}' instead.{}",
        key,
        file.display(),
        replacement,
        info.map(|i| format!("\n  {}", i)).unwrap_or_default()
    );
}
