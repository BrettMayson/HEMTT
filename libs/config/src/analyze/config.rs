use std::collections::{HashMap, HashSet};

use hemtt_common::reporting::{Code, Processed};
use hemtt_project::ProjectConfig;

use crate::{Class, Config, Ident, Item, Property, Str, Value};

use super::{
    codes::{
        ce3_duplicate_property::DuplicateProperty, ce7_missing_parent::MissingParent,
        cw1_parent_case::ParentCase, cw2_magwell_missing_magazine::MagwellMissingMagazine,
    },
    Analyze,
};

impl Analyze for Config {
    fn valid(&self, project: Option<&ProjectConfig>) -> bool {
        self.0.iter().all(|p| p.valid(project))
    }

    fn warnings(
        &self,
        project: Option<&ProjectConfig>,
        processed: &Processed,
    ) -> Vec<Box<dyn Code>> {
        let mut warnings = self
            .0
            .iter()
            .flat_map(|p| p.warnings(project, processed))
            .collect::<Vec<_>>();
        let mut defined = HashMap::new();
        warnings.extend(external_parent_case_warn(&self.0, &mut defined));
        if let Some(project) = project {
            warnings.extend(magwell_missing_magazine(project, self));
        }
        warnings
    }

    fn errors(&self, project: Option<&ProjectConfig>, processed: &Processed) -> Vec<Box<dyn Code>> {
        let mut errors = self
            .0
            .iter()
            .flat_map(|p| p.errors(project, processed))
            .collect::<Vec<_>>();
        let mut defined = HashSet::new();
        errors.extend(external_missing_error(&self.0, &mut defined));
        errors.extend(duplicate_properties(&self.0));
        errors
    }
}

fn external_missing_error(
    properties: &[Property],
    defined: &mut HashSet<String>,
) -> Vec<Box<dyn Code>> {
    let mut errors: Vec<Box<dyn Code>> = Vec::new();
    for property in properties {
        if let Property::Class(c) = property {
            match c {
                Class::External { name } => {
                    let name = name.value.to_lowercase();
                    if !defined.contains(&name) {
                        defined.insert(name);
                    }
                }
                Class::Local {
                    name,
                    parent,
                    properties,
                } => {
                    let name = name.value.to_lowercase();
                    if let Some(parent) = parent {
                        let parent = parent.value.to_lowercase();
                        if parent != name && !defined.contains(&parent) {
                            errors.push(Box::new(MissingParent::new(c.clone())));
                        }
                    }
                    defined.insert(name);
                    errors.extend(external_missing_error(properties, defined));
                }
            }
        }
    }
    errors
}

fn external_parent_case_warn(
    properties: &[Property],
    defined: &mut HashMap<String, Class>,
) -> Vec<Box<dyn Code>> {
    let mut warnings: Vec<Box<dyn Code>> = Vec::new();
    for property in properties {
        if let Property::Class(c) = property {
            match c {
                Class::External { name } => {
                    let name = name.value.to_lowercase();
                    defined.entry(name).or_insert_with(|| c.clone());
                }
                Class::Local {
                    name,
                    parent,
                    properties,
                } => {
                    let name_lower = name.value.to_lowercase();
                    if let Some(parent) = parent {
                        let parent_lower = parent.value.to_lowercase();
                        if parent_lower != name_lower {
                            if let Some(parent_class) = defined.get(&parent_lower) {
                                if parent_class.name().value != parent.value {
                                    warnings.push(Box::new(ParentCase::new(
                                        c.clone(),
                                        parent_class.clone(),
                                    )));
                                }
                            }
                        } else if parent.value != name.value {
                            warnings.push(Box::new(ParentCase::new(c.clone(), c.clone())));
                        }
                    }
                    defined.insert(name_lower, c.clone());
                    warnings.extend(external_parent_case_warn(properties, defined));
                }
            }
        }
    }
    warnings
}

fn duplicate_properties(properties: &[Property]) -> Vec<Box<dyn Code>> {
    let mut seen: HashMap<String, Vec<(bool, Ident)>> = HashMap::new();
    duplicate_properties_inner("", properties, &mut seen);
    let mut errors: Vec<Box<dyn Code>> = Vec::new();
    for (_, idents) in seen {
        if idents.len() > 1 && !idents.iter().all(|(class, _)| *class) {
            errors.push(Box::new(DuplicateProperty::new(
                idents.into_iter().map(|(_, i)| i).collect(),
            )));
        }
    }
    errors
}

fn duplicate_properties_inner(
    scope: &str,
    properties: &[Property],
    seen: &mut HashMap<String, Vec<(bool, Ident)>>,
) {
    for property in properties {
        match property {
            Property::Class(Class::Local {
                name, properties, ..
            }) => {
                duplicate_properties_inner(
                    &format!("{}.{}", scope, name.value.to_lowercase()),
                    properties,
                    seen,
                );
                let entry = seen
                    .entry(format!("{}.{}", scope, name.value.to_lowercase()))
                    .or_default();
                entry.push((true, name.clone()));
            }
            Property::Entry { name, .. } | Property::MissingSemicolon(name, _) => {
                let entry = seen
                    .entry(format!("{}.{}", scope, name.value.to_lowercase()))
                    .or_default();
                entry.push((false, name.clone()));
            }
            _ => (),
        }
    }
}

fn magwell_missing_magazine(project: &ProjectConfig, config: &Config) -> Vec<Box<dyn Code>> {
    let mut warnings: Vec<Box<dyn Code>> = Vec::new();
    let mut classes = Vec::new();
    let Some(Property::Class(Class::Local {
        properties: magwells,
        ..
    })) = config
        .0
        .iter()
        .find(|p| p.name().value.to_lowercase() == "cfgmagazinewells")
    else {
        return warnings;
    };
    let Some(Property::Class(Class::Local {
        properties: magazines,
        ..
    })) = config
        .0
        .iter()
        .find(|p| p.name().value.to_lowercase() == "cfgmagazines")
    else {
        return warnings;
    };
    for property in magazines {
        if let Property::Class(Class::Local { name, .. }) = property {
            classes.push(name);
        }
    }
    for magwell in magwells {
        let Property::Class(Class::Local {
            properties: addons, ..
        }) = magwell
        else {
            continue;
        };
        for addon in addons {
            let Property::Entry {
                name,
                value: Value::Array(magazines),
                ..
            } = addon
            else {
                continue;
            };
            for mag in &magazines.items {
                let Item::Str(Str { value, span }) = mag else {
                    continue;
                };
                if !value
                    .to_lowercase()
                    .starts_with(&project.prefix().to_lowercase())
                {
                    continue;
                }
                if !classes.iter().any(|c| c.value == *value) {
                    warnings.push(Box::new(MagwellMissingMagazine::new(
                        name.clone(),
                        value.clone(),
                        span.clone(),
                    )));
                }
            }
        }
    }
    warnings
}
