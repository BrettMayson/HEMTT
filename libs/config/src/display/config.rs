use crate::Config;

impl std::fmt::Display for Config {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for property in &self.0 {
            write!(f, "{property}")?;
        }
        Ok(())
    }
}
