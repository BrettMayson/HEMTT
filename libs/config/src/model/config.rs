use hemtt_common::version::Version;

use crate::{analyze::CfgPatch, Class, Number, Property, Value};

#[derive(Debug, PartialEq)]
/// A config file
pub struct Config(pub Vec<Property>);

impl Config {
    #[must_use]
    pub fn to_class(&self) -> Class {
        Class::Root {
            properties: self.0.clone(),
        }
    }
}

impl Config {
    #[must_use]
    /// Get the patches
    pub fn get_patches(&self) -> Vec<CfgPatch> {
        let mut patches = Vec::new();
        for property in &self.0 {
            if let Property::Class(Class::Local {
                name, properties, ..
            }) = property
            {
                if name.as_str() == "CfgPatches" {
                    for patch in properties {
                        if let Property::Class(Class::Local {
                            name, properties, ..
                        }) = patch
                        {
                            let mut required_version = Version::new(0, 0, 0, None);
                            for property in properties {
                                if let Property::Entry { name, value, .. } = property {
                                    if name.as_str() == "requiredVersion" {
                                        if let Value::Number(Number::Float32 { value, .. }) = value
                                        {
                                            required_version = Version::from(*value);
                                        }
                                    }
                                }
                            }
                            patches.push(CfgPatch::new(name.clone(), required_version));
                        }
                    }
                }
            }
        }
        patches
    }
}
