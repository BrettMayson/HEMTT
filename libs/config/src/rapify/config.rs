use std::io::Cursor;

use byteorder::{LittleEndian, WriteBytesExt};

use crate::{Class, Config};

use super::Rapify;

impl Rapify for Config {
    fn rapify<O: std::io::Write>(
        &self,
        output: &mut O,
        _offset: usize,
    ) -> Result<usize, std::io::Error> {
        output.write_all(b"\0raP")?;
        output.write_all(b"\0\0\0\0\x08\0\0\0")?;

        let root_class = Class::Root {
            properties: self.0.clone(),
        };
        let buffer: Box<[u8]> = vec![0; root_class.rapified_length()].into_boxed_slice();
        let mut cursor = Cursor::new(buffer);
        let written = root_class.rapify(&mut cursor, 16)?;
        assert_eq!(written, root_class.rapified_length());

        let enum_offset = 16 + cursor.get_ref().len() as u32;
        output.write_u32::<LittleEndian>(enum_offset)?;

        output.write_all(cursor.get_ref())?;

        output.write_all(b"\0\0\0\0")?;
        assert_eq!(written + 20, self.rapified_length());
        Ok(written + 20)
    }

    fn rapified_length(&self) -> usize {
        let root_class = Class::Root {
            properties: self.0.clone(),
        };
        root_class.rapified_length() + 20 // metadata
    }
}
