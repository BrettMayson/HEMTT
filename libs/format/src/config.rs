use logos::Logos;

use crate::FormatterConfig;

/// Check if we're in a context where minus should be treated as a negative number in config files
fn is_negative_number_context_config(output: &str) -> bool {
    output.ends_with('{')
        || output.ends_with('[')
        || output.ends_with('(')
        || output.ends_with("= ") || output.ends_with('=') || output.ends_with("= -")
        || output.ends_with(", ") || output.ends_with(", -")
        // Arithmetic operators - negative numbers can appear after them
        || output.ends_with("+ ") || output.ends_with("+ -")
        || output.ends_with("- ") || output.ends_with("- -")
        || output.ends_with("* ") || output.ends_with("* -")
        || output.ends_with("/ ") || output.ends_with("/ -")
        || output.ends_with("% ") || output.ends_with("% -")
}

/// Format nested braces with proper spacing
fn format_nested_braces(content: &str) -> String {
    if content.starts_with('{') && content.ends_with('}') {
        let inner = &content[1..content.len() - 1];
        let parts: Vec<&str> = inner.split(',').map(str::trim).collect();
        format!("{{{}}}", parts.join(", "))
    } else {
        content.to_string()
    }
}

/// Parse config array items from content
fn parse_config_array_items(content: &str) -> Vec<String> {
    let mut items = Vec::new();
    let mut current_item = String::new();
    let mut brace_depth = 0;
    let mut string_delimiter = None;
    let mut chars = content.chars().peekable();

    while let Some(ch) = chars.next() {
        match ch {
            '"' if string_delimiter.is_none() => {
                current_item.push(ch);
                string_delimiter = Some(ch);
            }
            '"' if Some(ch) == string_delimiter => {
                current_item.push(ch);
                // Check if this is an escaped quote (doubled)
                if chars.peek() == Some(&ch) {
                    if let Some(escaped_ch) = chars.next() {
                        current_item.push(escaped_ch);
                    }
                } else {
                    string_delimiter = None;
                }
            }
            '{' if string_delimiter.is_none() => {
                current_item.push(ch);
                brace_depth += 1;
            }
            '}' if string_delimiter.is_none() => {
                current_item.push(ch);
                brace_depth -= 1;
            }
            ',' if string_delimiter.is_none() && brace_depth == 0 => {
                let trimmed = current_item.trim();
                if !trimmed.is_empty() {
                    // Format nested braces with proper spacing
                    let formatted = if trimmed.starts_with('{') && trimmed.ends_with('}') {
                        format_nested_braces(trimmed)
                    } else {
                        trimmed.to_string()
                    };
                    items.push(formatted);
                }
                current_item.clear();
            }
            _ => {
                current_item.push(ch);
            }
        }
    }

    // Add the last item
    let trimmed = current_item.trim();
    if !trimmed.is_empty() {
        // Format nested braces with proper spacing
        let formatted = if trimmed.starts_with('{') && trimmed.ends_with('}') {
            format_nested_braces(trimmed)
        } else {
            trimmed.to_string()
        };
        items.push(formatted);
    }

    items
}

/// Format config array content with intelligent line breaking
fn format_config_array_content(
    content: &str,
    cfg: &FormatterConfig,
    indent_level: usize,
) -> String {
    let items = parse_config_array_items(content);

    if items.is_empty() {
        return String::new();
    }

    // Calculate one-line format
    let one_line = items.join(", ");
    let line_length = one_line.len();

    // Use single-line format for short arrays
    if line_length <= 100 {
        return one_line;
    }

    // Format with line breaks - try to fit multiple items per line like original
    let mut result = String::from("\n");
    result.push_str(&cfg.indent(indent_level + 1));

    let mut current_line_length = cfg.indent(indent_level + 1).len();
    let mut first_item = true;

    for item in items {
        let item_length = if first_item {
            item.len()
        } else {
            2 + item.len() // ", " + item
        };

        // Start a new line if this item would make the line too long
        if !first_item && current_line_length + item_length > 100 {
            result.push_str(",\n");
            result.push_str(&cfg.indent(indent_level + 1));
            result.push_str(&item);
            current_line_length = cfg.indent(indent_level + 1).len() + item.len();
        } else {
            if !first_item {
                result.push_str(", ");
            }
            result.push_str(&item);
            current_line_length += item_length;
        }

        first_item = false;
    }

    result.push('\n');
    result
}

#[derive(Logos, Debug, PartialEq)]
enum Token {
    // Keywords and identifiers
    #[regex(r"[A-Za-z0-9_]*")]
    Ident,

    // Numbers (integers, decimals, and hexadecimal)
    #[regex(r"0[xX][0-9a-fA-F]+|[0-9]+(\.[0-9]+)?")]
    Number,

    // Strings
    #[regex(r#""([^"\\]|\\.)*""#)]
    String,

    // Operators and punctuation
    #[token("=")]
    Eq,
    #[token(";")]
    Semi,
    #[token(",")]
    Comma,
    #[token(":")]
    Colon,
    #[token("{")]
    LBrace,
    #[token("}")]
    RBrace,
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("[")]
    LBracket,
    #[token("]")]
    RBracket,

    // Arithmetic operators
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("*")]
    Multiply,
    #[token("/")]
    Divide,
    #[token("%")]
    Modulo,

    // Preprocessor lines
    #[regex(r"#.*", priority = 2)]
    Preprocessor,

    // Comments
    #[regex(r"//[^\n]*")]
    LineComment,
    #[regex(r"/\*([^*]|\*[^/])*\*/")]
    BlockComment,

    // Whitespace tokens - we need to handle these to preserve empty lines
    #[regex(r"[ \t]+")]
    Space,
    #[regex(r"\n")]
    Newline,
}

#[allow(clippy::too_many_lines)]
/// Format a configuration file according to the provided `FormatterConfig`.
/// Returns the formatted string or an error message if formatting fails.
///
/// # Errors
/// Returns an error if the input contains unexpected tokens.
pub fn format_config(source: &str, cfg: &FormatterConfig) -> Result<String, String> {
    let mut lexer = Token::lexer(source);
    let mut output = String::new();
    let mut indent_level = 0;
    let mut need_indent = true;
    let mut in_array = false;
    let mut array_content = String::new();
    let mut brace_count = 0;
    let mut consecutive_newlines = 0;
    let mut after_paren = false;
    let mut newline_since_last_token = false;

    while let Some(token) = lexer.next() {
        let Ok(token) = token else {
            let span = lexer.span();
            let error_text = lexer.source().get(span.clone()).unwrap_or("unknown");
            return Err(format!(
                "Unexpected token '{}' at position {}..{}",
                error_text, span.start, span.end
            ));
        };

        match token {
            Token::Space => {
                // Skip spaces - we handle indentation ourselves
                consecutive_newlines = 0;
            }
            Token::Newline => {
                consecutive_newlines += 1;
                newline_since_last_token = true;
                // Preserve empty lines - when we see the second newline, it represents an empty line
                if consecutive_newlines == 1 {
                    // First newline - end the current line if not already ended
                    if !output.ends_with('\n') {
                        output.push('\n');
                    }
                } else if consecutive_newlines >= 2 {
                    // Additional newlines represent empty lines
                    output.push('\n');
                }
            }
            Token::Preprocessor => {
                consecutive_newlines = 0;
                if !output.is_empty() && !output.ends_with('\n') {
                    output.push('\n');
                }
                output.push_str(&cfg.indent(indent_level));
                output.push_str(lexer.slice().trim());
                output.push('\n');
                need_indent = true;
            }
            Token::LineComment => {
                consecutive_newlines = 0;

                if need_indent || newline_since_last_token {
                    // Comment at beginning of line or after a newline
                    if newline_since_last_token && !output.ends_with('\n') {
                        // Start a new line for the comment only if we're not already on a new line
                        output.push('\n');
                    }
                    output.push_str(&cfg.indent(indent_level));
                } else {
                    // True inline comment - on same line as previous content
                    output.push(' ');
                }
                output.push_str(lexer.slice().trim_end());
                output.push('\n');
                need_indent = true;
                newline_since_last_token = false;
            }
            Token::BlockComment => {
                consecutive_newlines = 0;
                let comment = lexer.slice();
                // Check if this is a multi-line comment
                let is_multiline = comment.contains('\n');

                if is_multiline {
                    // Multi-line block comment: treat like a line comment
                    if need_indent {
                        output.push_str(&cfg.indent(indent_level));
                    } else {
                        output.push(' ');
                    }
                    output.push_str(comment.trim_end());
                    output.push('\n');
                    need_indent = true;
                } else {
                    // Inline block comment: keep on same line
                    if !need_indent {
                        output.push(' ');
                    }
                    output.push_str(comment.trim());
                    need_indent = false;
                }
            }
            Token::LBracket => {
                consecutive_newlines = 0;
                // Don't add space before [ in array declarations
                output.push('[');
                need_indent = false;
            }
            Token::RBracket => {
                consecutive_newlines = 0;
                output.push(']');
                need_indent = false;
            }
            Token::LBrace => {
                consecutive_newlines = 0;
                if in_array {
                    brace_count += 1;
                    array_content.push('{');
                } else {
                    // Check if this is starting an array by looking ahead
                    let remaining = lexer.remainder();
                    if cfg.space_before_brace && !output.ends_with(' ') {
                        output.push(' ');
                    }
                    output.push('{');
                    if is_array_content(remaining) {
                        in_array = true;
                        array_content.clear();
                        brace_count = 0;
                    } else if is_empty_class(remaining) {
                        need_indent = false;
                    } else {
                        // Check if there's a comment on the same line (only whitespace before //)
                        let trimmed_remaining = remaining.trim_start();
                        if trimmed_remaining.starts_with("//") {
                            // There's a comment right after this brace on the same line
                            // Don't add newline, let comment handler manage the line ending
                            indent_level += 1;
                            need_indent = false;
                        } else {
                            // No immediate comment, proceed normally
                            output.push('\n');
                            indent_level += 1;
                            need_indent = true;
                        }
                    }
                }
            }
            Token::RBrace => {
                consecutive_newlines = 0;
                if in_array {
                    if brace_count > 0 {
                        brace_count -= 1;
                        array_content.push('}');
                    } else {
                        // End of array, format using local array function
                        let formatted =
                            format_config_array_content(&array_content, cfg, indent_level);

                        output.push_str(&formatted);
                        output.push('}');
                        in_array = false;
                        need_indent = false;
                    }
                } else {
                    // Check if this is an empty class (no newline before the brace)
                    if output.ends_with('{') {
                        // This is an empty class like "class Name {}"
                        output.push('}');
                        need_indent = false;
                    } else {
                        // Normal class ending
                        indent_level = indent_level.saturating_sub(1);
                        output.push_str(&cfg.indent(indent_level));
                        output.push('}');
                        output.push('\n');
                        need_indent = true;
                    }
                }
            }
            Token::Semi => {
                consecutive_newlines = 0;
                newline_since_last_token = false;
                if in_array {
                    // Don't add semicolons to array content, they are handled in formatting
                } else {
                    // If the last character is '}', put semicolon on same line
                    if output.trim_end().ends_with('}') {
                        // Remove the trailing newline from the previous RBrace
                        let trimmed = output.trim_end();
                        output.truncate(trimmed.len());
                    }
                    output.push(';');

                    // Look ahead to see if there's a comment on the same line
                    let remaining = lexer.remainder();
                    let next_line_comment = remaining.trim_start().starts_with("//");

                    if next_line_comment {
                        // Don't add newline, let the comment handler decide
                        need_indent = false;
                    } else {
                        output.push('\n');
                        need_indent = true;
                    }
                    after_paren = false;
                }
            }
            Token::Comma => {
                consecutive_newlines = 0;
                if in_array {
                    array_content.push(',');
                } else {
                    output.push(',');

                    // Check if there's already a space in the source after this comma
                    let span = lexer.span();
                    let source_after = &source[span.end..];

                    // Only add space if there's already one in the source (preserve original spacing)
                    if source_after.starts_with(' ') || source_after.starts_with('\t') {
                        output.push(' ');
                    }

                    need_indent = false;
                    // Don't set after_paren = false here, we're still inside parentheses
                }
            }
            Token::Colon => {
                consecutive_newlines = 0;
                if in_array {
                    array_content.push(':');
                } else {
                    output.push(':');
                    need_indent = false;
                }
            }
            Token::LParen => {
                consecutive_newlines = 0;
                if in_array {
                    array_content.push('(');
                } else {
                    output.push('(');
                    need_indent = false;
                    after_paren = true;
                }
            }
            Token::RParen => {
                consecutive_newlines = 0;
                if in_array {
                    array_content.push(')');
                } else {
                    output.push(')');
                    need_indent = false;
                    after_paren = false;
                }
            }
            // Arithmetic operators - add spaces around them, except for negative numbers
            Token::Plus | Token::Multiply | Token::Divide | Token::Modulo => {
                consecutive_newlines = 0;
                if in_array {
                    array_content.push_str(lexer.slice());
                } else {
                    if need_indent {
                        output.push_str(&cfg.indent(indent_level));
                        need_indent = false;
                    } else if !output.ends_with(' ') {
                        // Add space before operator
                        output.push(' ');
                    }
                    output.push_str(lexer.slice());
                    output.push(' '); // Add space after operator
                    newline_since_last_token = false;
                    after_paren = false;
                }
            }
            Token::Minus => {
                consecutive_newlines = 0;
                if in_array {
                    array_content.push_str(lexer.slice());
                } else {
                    if need_indent {
                        output.push_str(&cfg.indent(indent_level));
                        need_indent = false;
                    } else if is_negative_number_context_config(&output) {
                        // Negative number - no space before minus
                    } else if !output.ends_with(' ') {
                        // Arithmetic operation - add space before
                        output.push(' ');
                    }
                    output.push_str(lexer.slice());

                    // Check if this should have space after (arithmetic) or not (negative number)
                    let remaining = lexer.remainder().trim_start();
                    if remaining.chars().next().is_some_and(|c| c.is_ascii_digit()) {
                        // Next token is a digit - likely negative number, no space after
                    } else {
                        // Arithmetic operation - add space after
                        output.push(' ');
                    }
                    newline_since_last_token = false;
                    after_paren = false;
                }
            }
            Token::Eq => {
                consecutive_newlines = 0;
                if in_array {
                    array_content.push_str(lexer.slice());
                } else {
                    if need_indent {
                        output.push_str(&cfg.indent(indent_level));
                        need_indent = false;
                    } else if !output.ends_with(' ') {
                        // Add space before equals
                        output.push(' ');
                    }
                    output.push_str(lexer.slice());
                    output.push(' '); // Always add space after equals
                    newline_since_last_token = false;
                    after_paren = false;
                }
            }
            _ => {
                consecutive_newlines = 0;
                if in_array {
                    array_content.push_str(lexer.slice());
                } else {
                    if need_indent {
                        output.push_str(&cfg.indent(indent_level));
                        need_indent = false;
                    } else if !after_paren && !matches!(token, Token::LBracket | Token::RBracket) {
                        // Check if we should add space before this token
                        let ends_with_minus = output.ends_with('-');
                        let is_number = matches!(token, Token::Number);
                        let is_negative_number_context = ends_with_minus
                            && is_number
                            && is_negative_number_context_config(&output);

                        let should_add_space = !(output.ends_with(' ')
                            || output.ends_with('(')
                            || output.ends_with('[')
                            || output.ends_with('=')
                            || output.ends_with('{')
                            || output.ends_with(',')
                            // Special case: don't add space after minus when followed by a number (negative number)
                            || is_negative_number_context
                            // Don't add space before braces - let LBrace handler manage spacing
                            || matches!(token, Token::LBrace));

                        if should_add_space {
                            output.push(' ');
                        }
                    }
                    output.push_str(lexer.slice());
                    newline_since_last_token = false;
                    // Don't automatically set after_paren = false here
                    // Let it be managed by specific token handlers
                }
            }
        }
    }

    Ok(output.trim_end().to_string() + "\n")
}

fn is_array_content(content: &str) -> bool {
    // Heuristic: if it contains mostly numbers, commas, and braces, it's likely an array
    let trimmed = content.trim();
    if trimmed.is_empty() {
        return false;
    }

    // Find the content up to the first unmatched closing brace
    let mut brace_depth = 0;
    let mut array_end = None;
    let chars = trimmed.char_indices();

    for (i, ch) in chars {
        match ch {
            '{' => brace_depth += 1,
            '}' => {
                if brace_depth == 0 {
                    array_end = Some(i);
                    break;
                }
                brace_depth -= 1;
            }
            _ => {}
        }
    }

    // Get just the content within the braces
    let array_content = if let Some(end) = array_end {
        &trimmed[..end]
    } else {
        return false; // No closing brace found
    };

    if array_content.is_empty() {
        return false;
    }

    // Check if the content looks like an array (numbers, letters for macros, commas, braces, quotes, parentheses, decimal points)
    let chars = array_content.chars();
    let mut has_content = false;

    for ch in chars {
        match ch {
            'A'..='Z' | 'a'..='z' | '0'..='9' | '_' | '{' | '}' | '"' | '(' | ')' | '.' => {
                has_content = true;
            }
            ',' | ' ' | '\t' | '\n' => {
                // These are fine
            }
            _ => return false, // Something else, not an array
        }
    }

    has_content
}

fn is_empty_class(content: &str) -> bool {
    // Check if the content is empty (only whitespace before closing brace)
    let trimmed = content.trim();
    trimmed.starts_with('}')
}
