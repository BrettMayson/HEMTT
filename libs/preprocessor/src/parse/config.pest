file = _{ SOI ~ (
    word | alpha | digit | underscore | dash | assignment | plus | left_brace | right_brace | left_bracket | right_bracket |
    left_parentheses | right_parentheses | colon | semicolon | join | directive | escape | slash | comma | decimal |
    double_quote | single_quote | left_angle | right_angle | unicode | newline
)* ~ EOI }
word = ${ ( underscore | alpha ) ~ ( alpha | digit | underscore )* }
alpha = { LETTER }
digit = { ASCII_DIGIT }
underscore = { "_" }
dash = { "-" }
assignment = { "=" }
plus = { "+" }
left_brace = { "{" }
right_brace = { "}" }
left_bracket = { "[" }
right_bracket = { "]" }
left_parentheses = { "(" }
right_parentheses = { ")" }
colon = { ":" }
semicolon = { ";" }
join = { "##" }
directive = { "#" }
escape = { "\\" }
slash = { "/" }
comma = { "," }
decimal = { "." }
double_quote = { "\"" }
single_quote = { "'" }
left_angle = { "<" }
right_angle = { ">" }
unicode = { LETTER | MARK | NUMBER | SYMBOL | PUNCTUATION | MATH | OTHER_MATH }
newline = { NEWLINE }

space = { " " }
tab = { "\t" }
WHITESPACE = ${ space | tab }
COMMENT = { ("/*" ~ (!"*/" ~ ANY)* ~ "*/") }
