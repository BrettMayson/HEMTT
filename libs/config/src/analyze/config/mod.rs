use std::sync::Arc;

use hemtt_common::config::ProjectConfig;
use hemtt_workspace::reporting::{Code, Processed};

use crate::Config;

use super::Analyze;

mod duplicate_classes;
mod duplicate_properties;
mod external_missing;
mod external_parent_case;
mod magwells;

impl Analyze for Config {
    fn warnings(
        &self,
        project: Option<&ProjectConfig>,
        processed: &Processed,
    ) -> Vec<Arc<dyn Code>> {
        let mut warnings = self
            .0
            .iter()
            .flat_map(|p| p.warnings(project, processed))
            .collect::<Vec<_>>();
        warnings.extend(external_parent_case::warn(&self.0, processed));
        if let Some(project) = project {
            warnings.extend(magwells::missing_magazine(project, self, processed));
        }
        warnings
    }

    fn errors(&self, project: Option<&ProjectConfig>, processed: &Processed) -> Vec<Arc<dyn Code>> {
        let mut errors = self
            .0
            .iter()
            .flat_map(|p| p.errors(project, processed))
            .collect::<Vec<_>>();
        errors.extend(duplicate_classes::error(&self.0, processed));
        errors.extend(external_missing::error(&self.0, processed));
        errors.extend(duplicate_properties::duplicate_properties(
            &self.0, processed,
        ));
        errors
    }
}
