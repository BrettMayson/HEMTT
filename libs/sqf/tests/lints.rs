#![allow(clippy::unwrap_used)]

use std::{io::Read, sync::Arc};

use hemtt_common::config::ProjectConfig;
use hemtt_preprocessor::Processor;
use hemtt_sqf::{analyze::analyze, parser::database::Database};
use hemtt_workspace::{addons::Addon, reporting::WorkspaceFiles, LayerType};

const ROOT: &str = "tests/lints/";

macro_rules! analyze {
    ($dir:ident) => {
        paste::paste! {
            #[test]
            fn [<simple_ $dir>]() {
                test_analyze(stringify!($dir));
            }
        }
    };
}

fn test_analyze(dir: &str) {
    let folder = std::path::PathBuf::from(ROOT).join(dir);
    let workspace = hemtt_workspace::Workspace::builder()
        .physical(&folder, LayerType::Source)
        .finish(None, false, &hemtt_common::config::PDriveOption::Disallow)
        .unwrap();
    let source = workspace.join("source.sqf").unwrap();
    let processed = Processor::run(&source).unwrap();
    let database = Arc::new(Database::a3(false));
    let workspace_files = WorkspaceFiles::new();

    let config_path_full = std::path::PathBuf::from(ROOT).join("project_tests.toml");
    let config = ProjectConfig::from_file(&config_path_full).unwrap();

    match hemtt_sqf::parser::run(&database, &processed) {
        Ok(sqf) => {
            let codes = analyze(
                &sqf,
                Some(&config),
                &processed,
                Arc::new(Addon::test_addon()),
                database.clone(),
            );
            let stdout = codes
                .iter()
                .map(|e| e.diagnostic().unwrap().to_string(&workspace_files))
                .collect::<Vec<_>>()
                .join("\n")
                .replace('\r', "");
            let mut expected = Vec::new();
            std::fs::File::open(folder.join("stdout.ansi"))
                .unwrap()
                .read_to_end(&mut expected)
                .unwrap();
            if expected.is_empty() {
                std::fs::write(folder.join("stdout.ansi"), stdout.as_bytes()).unwrap();
            }
            let expected = String::from_utf8_lossy(&expected).replace('\r', "");
            assert_eq!(stdout, expected);
        }
        Err(hemtt_sqf::parser::ParserError::ParsingError(e)) => {
            for error in e {
                println!(
                    "{}",
                    error.diagnostic().unwrap().to_string(&workspace_files)
                );
            }
            panic!("failed to parse");
        }
        Err(e) => panic!("{e:?}"),
    };
}

analyze!(s02_event_handler_case);
analyze!(s03_static_typename);
analyze!(s04_command_case);
analyze!(s05_if_assign);
analyze!(s06_find_in_str);
analyze!(s07_select_parse_number);
analyze!(s08_format_args);
analyze!(s09_banned_command);
analyze!(s11_if_not_else);
analyze!(s17_var_all_caps);
analyze!(s18_in_vehicle_check);
analyze!(s20_bool_static_comparison);
analyze!(s21_invalid_comparisons);
analyze!(s22_this_call);
analyze!(s23_reassign_reserved_variable);
