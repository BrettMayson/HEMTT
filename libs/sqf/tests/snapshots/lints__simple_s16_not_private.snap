---
source: libs/sqf/tests/lints.rs
expression: "lint(stringify! (s16_not_private), false).0"
---
[0m[1m[38;5;14mhelp[L-S16][0m[1m: Local variable `_z` is not private[0m
  [0m[36mâ”Œâ”€[0m s16_not_private.sqf:1:1
  [0m[36mâ”‚[0m
[0m[36m1[0m [0m[36mâ”‚[0m [0m[36m_z[0m = 6;
  [0m[36mâ”‚[0m [0m[36m^^[0m [0m[36mcan be private[0m
  [0m[36mâ”‚[0m
  [0m[36m=[0m [32mtry[0m: private _z
